File: quotation/core/integrated.py
"""레노버 x86 구성 파일(통합 모드)의 해석 규칙.

문서 형식은 IBM eConfig Export 와 같다. 다른 것은 **값이 뜻하는 바** 이며,
지금까지 확인된 차이는 둘뿐이다. 두 가지 모두 여기 모아 둔다.

1. 장비 이름이 ProductDescription 이 아니라 ProductName 에 있다.

   한 파일에 SR650 V4 를 열한 대 담으면 열 그룹의 ProductDescription 이 전부
   'ThinkSystem SR650 V4-3yr Base Warranty' 로 같다. 사람이 붙인 이름
   ('백업서버_1식', '웹서버_1식')은 ProductName 에 있고, 손으로 만든 견적서의
   종목 칸도 그 이름을 쓴다.

2. 본체 라인의 UnitListPrice 가 **그 서버 한 대의 전체 LP** 다.

   Lenovo DCSC 가 내려 주는 요약표(Summary)와 대조해서 확인했다. 예를 들어
   그룹 1000 은 요약표에서 다음처럼 갈라지는데,

       ThinkSystem SR650 V4      27,360,001.8
       Red Hat RHEL               5,440,001
       XClarity Controller          840,001
       3Yr Premier 24x7 4Hr       5,691,000
       3Yr KYD Add-On               841,000
       Configuration Instruction          1
                                 ------------
                                 40,172,001.8   <- XML 의 본체 라인 UnitListPrice

   XML 의 소프트웨어 라인은 값이 1(자리표)일 뿐 실제 금액을 담지 않는다.
   그래서 소프트웨어·서비스 라인의 금액을 그대로 더하면 서비스 금액이 두 번
   들어가고 자리표 1원까지 얹힌다.

   갈라 놓을 값은 문서 안에 있다. `SectionData/GroupData` 의 요약표가 품목별
   실금액을 담고 있어서, 소프트웨어·서비스에 제 금액을 적고 하드웨어는 남는
   값으로 잡는다 (`dcsc_summary.py`). 요약표를 못 읽으면 지금까지대로 본체
   라인 하나로만 센다.

   어느 쪽이든 **장비군 합계는 본체 LP 그대로** 이며, 라인은 지우지 않는다.
   무엇이 들어 있는지는 상세 시트에 그대로 남는다.
"""
from __future__ import annotations

from dataclasses import replace
from decimal import Decimal

from . import dcsc_summary
from .models import LineItem
from .money import is_priced, to_decimal
from .naming import product_key


def body_line(items: tuple[LineItem, ...]) -> LineItem | None:
    """그룹의 본체 라인. CPUSIUvalue=1 인 하드웨어 라인이다."""
    return next((i for i in items if i.siu == 1 and i.is_hardware), None)


def group_key(items: tuple[LineItem, ...]) -> str:
    """장비군 이름. 본체 라인의 ProductName 을 먼저 쓴다."""
    line = body_line(items) or items[0]
    return product_key(line.product_name, line.description)


def fold_prices(items: tuple[LineItem, ...],
                summary: dcsc_summary.Summary | None = None,
                ) -> tuple[LineItem, ...]:
    """본체 LP 에 이미 들어 있는 금액을 갈라 놓는다 (머리말 2번).

    값이 붙은 본체 라인이 있을 때만 손댄다. 랙·PDU·케이블처럼 본체 없이
    부품만 들어 있는 그룹은 저마다 제 금액을 가지므로 그대로 둔다.

    Args:
        items: 장비군 한 개의 라인.
        summary: 문서의 요약표. 주면 품목별 실금액을 적고, 없거나 이 장비군에
            맞는 구성이 없으면 하드웨어가 아닌 라인의 금액 칸을 비운다.
    """
    body = body_line(items)
    if body is None or not is_priced(body.unit_price):
        return items
    config = summary.take(body.part_number, body.unit_price) if summary else None
    if config is not None:
        split = _split_prices(items, body, config)
        if split is not None:
            return split
    return tuple(item if item.is_hardware else replace(item, unit_price=None)
                 for item in items)


def _split_prices(items: tuple[LineItem, ...], body: LineItem,
                  config: dcsc_summary.Config) -> tuple[LineItem, ...] | None:
    """요약표의 실금액을 라인에 적고 하드웨어는 남는 값으로 잡는다.

    남는 값에 자리표 몇 원과 소수 넷째 자리 반올림이 모이므로 장비군 합계는
    본체 LP 와 한 푼도 어긋나지 않는다.

    같은 품번이 그 장비군에 두 번 나오면 앞선 라인이 금액을 가져가고 뒤는
    비운다. 합계는 어느 쪽이든 같다.

    Returns:
        고쳐 쓴 라인. 남는 값이 음수면 짝을 잘못 지은 것이므로 None.
    """
    left = dict(config.prices)
    priced: list[LineItem] = []
    others = Decimal(0)
    for item in items:
        if item.is_hardware:
            priced.append(item)
            continue
        price = left.pop(dcsc_summary.part_key(item.part_number), None)
        if not price:  # 요약표에 없거나 자리표뿐인 라인 (Configuration Instruction)
            priced.append(replace(item, unit_price=None))
            continue
        others += price
        priced.append(replace(item, unit_price=price))

    rest = to_decimal(body.unit_price) - others
    if rest < 0:
        return None
    return tuple(replace(item, unit_price=rest) if item is body else item
                 for item in priced)

File: rust/core/src/integrated.rs
//! 레노버 x86 구성 파일(통합 모드)의 해석 규칙.
//! 파이썬 `quotation/core/integrated.py`.
//!
//! 문서 형식은 IBM eConfig Export 와 같다. 다른 것은 **값이 뜻하는 바** 이며,
//! 지금까지 확인된 차이는 둘뿐이다.
//!
//! 1. 장비 이름이 ProductDescription 이 아니라 ProductName 에 있다.
//!    한 파일에 같은 기종을 여러 대 담으면 설명이 전부 같고, 사람이 붙인 이름
//!    (`백업서버_1식`)만 장비를 구분한다.
//! 2. 본체 라인의 UnitListPrice 가 **그 서버 한 대의 전체 LP** 다. XML 의
//!    소프트웨어 라인은 자리표 1 원일 뿐이라 그대로 더하면 두 번 센다.
//!    요약표(`dcsc_summary`)를 읽으면 품목별 실금액을 적고, 못 읽으면 금액
//!    칸을 비운다. **어느 쪽이든 장비군 합계는 본체 LP 그대로다.**

use crate::dcsc_summary::{self, Summary};
use crate::models::LineItem;
use crate::money::{Amount, Decimal, is_priced, to_decimal};
use crate::naming::product_key;

/// 그룹의 본체 라인. CPUSIUvalue=1 인 하드웨어 라인이다.
pub fn body_index(items: &[LineItem]) -> Option<usize> {
    items
        .iter()
        .position(|item| item.siu == 1 && item.is_hardware())
}

/// 장비군 이름. 본체 라인의 ProductName 을 먼저 쓴다.
pub fn group_key(items: &[LineItem]) -> String {
    let line = body_index(items).map_or(&items[0], |index| &items[index]);
    product_key(&line.product_name, &line.description)
}

/// 본체 LP 에 이미 들어 있는 금액을 갈라 놓는다 (머리말 2번).
///
/// 값이 붙은 본체 라인이 있을 때만 손댄다. 랙·PDU·케이블처럼 본체 없이
/// 부품만 들어 있는 그룹은 저마다 제 금액을 가지므로 그대로 둔다.
pub fn fold_prices(items: Vec<LineItem>, summary: Option<&mut Summary>) -> Vec<LineItem> {
    let Some(body) = body_index(&items) else {
        return items;
    };
    if !is_priced(&items[body].unit_price) {
        return items;
    }

    let config =
        summary.and_then(|summary| summary.take(&items[body].part_number, &items[body].unit_price));
    if let Some(mut config) = config {
        if let Some(split) = split_prices(&items, body, &mut config) {
            return split;
        }
    }
    // 요약표가 없으면 하드웨어가 아닌 라인의 금액 칸을 비운다.
    items
        .into_iter()
        .map(|item| {
            if item.is_hardware() {
                item
            } else {
                LineItem {
                    unit_price: Amount::Missing,
                    ..item
                }
            }
        })
        .collect()
}

/// 요약표의 실금액을 라인에 적고 하드웨어는 남는 값으로 잡는다.
///
/// 남는 값에 자리표 몇 원과 소수 넷째 자리 반올림이 모이므로 장비군 합계는
/// 본체 LP 와 한 푼도 어긋나지 않는다.
///
/// 같은 품번이 그 장비군에 두 번 나오면 앞선 라인이 금액을 가져가고 뒤는
/// 비운다. 합계는 어느 쪽이든 같다.
///
/// 남는 값이 음수면 짝을 잘못 지은 것이므로 `None`.
fn split_prices(
    items: &[LineItem],
    body: usize,
    config: &mut dcsc_summary::Config,
) -> Option<Vec<LineItem>> {
    let mut priced: Vec<LineItem> = Vec::with_capacity(items.len());
    let mut others = Decimal::ZERO;
    for item in items {
        if item.is_hardware() {
            priced.push(item.clone());
            continue;
        }
        match config.take_price(&dcsc_summary::part_key(&item.part_number)) {
            // 요약표에 없거나 자리표뿐인 라인 (Configuration Instruction).
            None => priced.push(LineItem {
                unit_price: Amount::Missing,
                ..item.clone()
            }),
            Some(price) if price.is_zero() => priced.push(LineItem {
                unit_price: Amount::Missing,
                ..item.clone()
            }),
            Some(price) => {
                others += price;
                priced.push(LineItem {
                    unit_price: Amount::Priced(price),
                    ..item.clone()
                });
            }
        }
    }

    let rest = to_decimal(&items[body].unit_price) - others;
    if rest.is_sign_negative() {
        return None;
    }
    priced[body].unit_price = Amount::Priced(rest);
    Some(priced)
}
